File: server-python/gemini.py
"""Gemini provider: wraps google-genai's explicit function/tool calling.

This is a behavior-preserving move of what used to be inline in
app.py (client construction, GENERATE_CONFIG, build_contents, and the
Gemini-specific parts of run_agent_loop) — nothing about how Gemini is
called has changed, it's just reachable through the Provider interface
now so agent.py doesn't need to import google.genai at all.
"""

import os

from google import genai
from google.genai import types

from prompts import SYSTEM_INSTRUCTION
from providers.base import Provider, ProviderCapabilities, ProviderResponse, ToolCall
from tools import TOOL_SCHEMAS

# tools.TOOL_SCHEMAS is written in standard (lowercase) JSON Schema so
# it can be shared as-is with OpenAICompatibleProvider. Gemini's
# FunctionDeclaration.parameters expects the same shape but with
# uppercase type names (OBJECT, STRING, ...), so this is the one
# Gemini-specific translation step.
_GEMINI_TYPE_MAP = {
    "object": "OBJECT",
    "string": "STRING",
    "boolean": "BOOLEAN",
    "integer": "INTEGER",
    "number": "NUMBER",
    "array": "ARRAY",
}


def _to_gemini_schema(schema: dict) -> dict:
    """Recursively uppercase JSON Schema `type` values for Gemini."""

    converted = dict(schema)

    if "type" in converted:
        converted["type"] = _GEMINI_TYPE_MAP.get(
            converted["type"], converted["type"]
        )

    if "properties" in converted:
        converted["properties"] = {
            key: _to_gemini_schema(value)
            for key, value in converted["properties"].items()
        }

    if "items" in converted:
        converted["items"] = _to_gemini_schema(converted["items"])

    return converted


def _build_tools() -> list:
    return [
        types.Tool(
            function_declarations=[
                types.FunctionDeclaration(
                    name=name,
                    description=schema["description"],
                    parameters=_to_gemini_schema(schema["parameters"]),
                )
                for name, schema in TOOL_SCHEMAS.items()
            ]
        )
    ]


def _sanitize_part(part: dict) -> types.Part | None:
    """Convert a history part dict into a types.Part the Gemini SDK accepts.

    The frontend stores conversation history as plain JSON, so parts arrive
    here as dicts. Passing raw dicts into types.Content causes pydantic
    validation errors on the SDK side ("Input should be a valid dictionary or
    object to extract fields from"). This mirrors the TypeScript
    gemini.ts sanitizeParts() logic: keep only the part shapes Gemini
    accepts (text, functionCall, functionResponse) and drop everything else.
    """
    if not isinstance(part, dict):
        return None

    if isinstance(part.get("text"), str) and part["text"]:
        return types.Part(text=part["text"])

    function_call = part.get("functionCall")
    if isinstance(function_call, dict) and function_call.get("name"):
        return types.Part(
            function_call=types.FunctionCall(
                name=function_call["name"],
                args=function_call.get("args") or {},
                id=function_call.get("id") or None,
            )
        )

    function_response = part.get("functionResponse")
    if isinstance(function_response, dict) and function_response.get("name"):
        return types.Part(
            function_response=types.FunctionResponse(
                name=function_response["name"],
                response=function_response.get("response"),
                id=function_response.get("id") or None,
            )
        )

    return None


class GeminiProvider(Provider):

    def __init__(self, api_key: str = None, model: str = None):
        # Prefer values supplied by the factory (ProviderConfig). Fall
        # back to the environment only when the caller omitted an
        # argument entirely, so direct construction still works without
        # going through get_provider().
        if api_key is None:
            api_key = os.getenv("GOOGLE_API_KEY")

        if not api_key:
            raise RuntimeError(
                "GOOGLE_API_KEY is not set. Add it to your .env file."
            )

        self.client = genai.Client(api_key=api_key)

        # Gemini 3.6 Flash is Google's current GA Flash model tuned for
        # coding, tool-use, and multi-step agentic workloads, so it's
        # the default here. Override with GEMINI_MODEL (or pass model=)
        # if you want a different currently supported Gemini model.
        if model is None:
            model = os.getenv("GEMINI_MODEL", "gemini-3.6-flash")

        self.model_name = model
        self.model = self.model_name
        self._capabilities = ProviderCapabilities(
            tools=True,
            streaming=True,
            model_listing=False,
            requires_api_key=True,
            local=False,
        )

        self.config = types.GenerateContentConfig(
            tools=_build_tools(),
            automatic_function_calling=types.AutomaticFunctionCallingConfig(
                disable=True
            ),
            system_instruction=SYSTEM_INSTRUCTION,
        )

    @property
    def capabilities(self) -> ProviderCapabilities:
        return self._capabilities

    def build_contents(self, msg, history):
        contents = []

        for item in history:
            role = item.get("role")
            parts = item.get("parts", [])

            if role and parts:
                converted_parts = []

                for part in parts:
                    sanitized = _sanitize_part(part)
                    if sanitized is not None:
                        converted_parts.append(sanitized)

                if converted_parts:
                    contents.append(
                        types.Content(role=role, parts=converted_parts)
                    )

        contents.append(
            types.Content(
                role="user",
                parts=[types.Part.from_text(text=msg)],
            )
        )

        return contents

    def generate(self, contents) -> ProviderResponse:
        response = self.client.models.generate_content(
            model=self.model_name,
            contents=contents,
            config=self.config,
        )

        if not response.candidates:
            raise RuntimeError("Gemini returned no response candidates.")

        model_content = response.candidates[0].content

        if model_content is None or not model_content.parts:
            raise RuntimeError("Gemini returned an empty response.")

        tool_calls = [
            ToolCall(
                name=part.function_call.name,
                args=dict(part.function_call.args or {}),
            )
            for part in model_content.parts
            if part.function_call
        ]

        text = None

        if not tool_calls:
            try:
                text = response.text
            except Exception:
                text = None

            if not text:
                # Fall back to manually concatenating any text parts.
                text = "".join(
                    part.text
                    for part in model_content.parts
                    if getattr(part, "text", None)
                ) or None

        return ProviderResponse(text=text, tool_calls=tool_calls, raw=model_content)

    def append_model_turn(self, contents, response):
        # Preserve Gemini's function-call (or text) message.
        # For direct command responses, response.raw is None — skip it
        # so the next provider.generate() call doesn't pass a None entry
        # to the Gemini SDK (which would fail pydantic validation).
        if response.raw is None:
            return contents
        return contents + [response.raw]

    def append_tool_results(self, contents, results):
        # IMPORTANT: google-genai 2.17.0 does not accept id= here.
        parts = [
            types.Part.from_function_response(
                name=item["name"],
                response={"result": item["result"]},
            )
            for item in results
        ]

        return contents + [types.Content(role="user", parts=parts)]

File: server-typescript/src/providers/gemini.ts
import { Provider, ProviderCapabilities, ProviderResponse, ToolCall } from "./base.ts";
import { CHAT_ONLY_INSTRUCTION, SYSTEM_INSTRUCTION } from "../prompts.ts";
import { buildToolSchemas } from "../tools.ts";

interface GeminiPart {
  text?: string;
  functionCall?: { name?: string; args?: Record<string, unknown>; id?: string };
  functionResponse?: { name?: string; response?: unknown; id?: string };
}

interface GeminiContent {
  role: "user" | "model";
  parts: GeminiPart[];
}

export class GeminiProvider extends Provider {
  readonly baseUrl: string;
  readonly apiKey: string | undefined;
  readonly timeout: number;
  readonly displayName = "Gemini";
  private readonly tools: unknown[];

  protected _capabilities: ProviderCapabilities = {
    tools: true,
    streaming: false,
    model_listing: true,
    requires_api_key: true,
    local: false,
    notes: "Native Gemini generateContent API; streaming is not implemented yet.",
  };

  constructor(config: {
    model: string;
    api_key?: string;
    base_url?: string;
    timeout?: number;
  }) {
    super();
    this.name = "gemini";
    this.model = config.model;
    this.apiKey = config.api_key;
    this.baseUrl = (config.base_url || "https://generativelanguage.googleapis.com/v1beta").replace(/\/$/, "");
    this.timeout = config.timeout ?? 120;
    this.tools = buildToolSchemas().map((tool) => ({
      functionDeclarations: [
        {
          name: tool.function.name,
          description: tool.function.description,
          parameters: tool.function.parameters,
        },
      ],
    }));
  }

  buildContents(msg: string, history: unknown[]): unknown[] {
    const contents: GeminiContent[] = [];
    for (const item of history) {
      const normalized = this.normalizeHistoryItem(item);
      if (normalized) contents.push(normalized);
    }
    contents.push({ role: "user", parts: [{ text: msg }] });
    return contents;
  }

  async generate(contents: unknown[]): Promise<ProviderResponse> {
    this.requireApiKey();
    const response = await this.request("POST", `${this.baseUrl}/models/${encodeURIComponent(this.model)}:generateContent?key=${encodeURIComponent(this.apiKey!)}`, {
      body: JSON.stringify({
        contents,
        systemInstruction: {
          role: "system",
          parts: [{ text: this.capabilities.tools ? SYSTEM_INSTRUCTION : CHAT_ONLY_INSTRUCTION }],
        },
        generationConfig: {},
        tools: this.capabilities.tools ? this.tools : undefined,
      }),
    });

    if (!response.ok) {
      throw new Error(await this.apiError(response, "Gemini request failed"));
    }

    const data = (await response.json()) as Record<string, unknown>;
    const candidate = ((data.candidates as unknown[]) || [])[0] as Record<string, unknown> | undefined;
    const content = candidate?.content as { parts?: GeminiPart[]; role?: string } | undefined;
    const parts = content?.parts || [];
    const toolCalls: ToolCall[] = [];
    for (const part of parts) {
      if (!part.functionCall?.name) continue;
      toolCalls.push({
        name: part.functionCall.name,
        args: part.functionCall.args || {},
        id: part.functionCall.id,
      });
    }

    const text = parts
      .map((part) => part.text || "")
      .filter(Boolean)
      .join("") || null;

    return { text, tool_calls: toolCalls, raw: content || null };
  }

  appendModelTurn(contents: unknown[], response: ProviderResponse): unknown[] {
    if (response.raw && typeof response.raw === "object") {
      return [...contents, response.raw];
    }
    return [...contents, { role: "model", parts: [{ text: response.text || "" }] }];
  }

  appendToolResults(
    contents: unknown[],
    results: { name: string; result: unknown }[]
  ): unknown[] {
    const previous = contents.at(-1) as GeminiContent | undefined;
    const calls = (previous?.parts || [])
      .map((part) => part.functionCall)
      .filter((call): call is NonNullable<GeminiPart["functionCall"]> => !!call?.name);

    const parts = results.map((result, index) => {
      const call = calls[index];
      if (!call) {
        throw new Error(`Gemini tool result '${result.name}' has no matching function call.`);
      }
      return {
        functionResponse: {
          name: result.name,
          response: { result: result.result },
          ...(call.id ? { id: call.id } : {}),
        },
      };
    });

    return [...contents, { role: "user", parts }];
  }

  async listModels(): Promise<unknown[]> {
    if (!this.apiKey) return [];
    try {
      const response = await this.request("GET", `${this.baseUrl}/models?key=${encodeURIComponent(this.apiKey)}`);
      if (!response.ok) return [];
      const data = (await response.json()) as { models?: unknown[] };
      return (data.models || [])
        .map((model) => {
          if (typeof model === "string") return { id: model.replace(/^models\//, "") };
          if (!model || typeof model !== "object") return null;
          const name = (model as Record<string, unknown>).name;
          return typeof name === "string" ? { id: name.replace(/^models\//, "") } : null;
        })
        .filter((model): model is { id: string } => model !== null);
    } catch {
      return [];
    }
  }

  async probe(): Promise<{ available: boolean; error: string | null }> {
    if (!this.apiKey) return { available: false, error: "Gemini API key is not configured (GOOGLE_API_KEY)." };
    try {
      const response = await this.request("GET", `${this.baseUrl}/models?key=${encodeURIComponent(this.apiKey)}`);
      if (response.ok) return { available: true, error: null };
      return { available: false, error: await this.apiError(response, "Gemini probe failed") };
    } catch (exc) {
      return { available: false, error: exc instanceof Error ? exc.message : String(exc) };
    }
  }

  /**
   * Map a client history turn into a Gemini Content object.
   *
   * The React client stores UI-only metadata (for example `toolActivity`) on
   * model turns. Gemini's generateContent schema only allows `role` and
   * `parts` on each content item, so any extra fields must be stripped or the
   * API returns HTTP 400 (Unknown name "toolActivity").
   */
  private normalizeHistoryItem(item: unknown): GeminiContent | null {
    if (!item || typeof item !== "object") return null;
    const record = item as Record<string, unknown>;

    if (record.role === "user" || record.role === "model") {
      if (Array.isArray(record.parts)) {
        const parts = this.sanitizeParts(record.parts);
        if (parts.length === 0) return null;
        return { role: record.role, parts };
      }
      if (typeof record.content === "string" && record.content) {
        return { role: record.role, parts: [{ text: record.content }] };
      }
    }

    if (record.role === "assistant" && typeof record.content === "string" && record.content) {
      return { role: "model", parts: [{ text: record.content }] };
    }

    return null;
  }

  /** Keep only part shapes Gemini accepts; drop unknown keys. */
  private sanitizeParts(parts: unknown[]): GeminiPart[] {
    const sanitized: GeminiPart[] = [];
    for (const part of parts) {
      if (!part || typeof part !== "object") continue;
      const record = part as Record<string, unknown>;

      if (typeof record.text === "string") {
        sanitized.push({ text: record.text });
        continue;
      }

      if (record.functionCall && typeof record.functionCall === "object") {
        const call = record.functionCall as Record<string, unknown>;
        if (typeof call.name === "string" && call.name) {
          sanitized.push({
            functionCall: {
              name: call.name,
              args: (call.args as Record<string, unknown>) || {},
              ...(typeof call.id === "string" ? { id: call.id } : {}),
            },
          });
        }
        continue;
      }

      if (record.functionResponse && typeof record.functionResponse === "object") {
        const resp = record.functionResponse as Record<string, unknown>;
        if (typeof resp.name === "string" && resp.name) {
          sanitized.push({
            functionResponse: {
              name: resp.name,
              response: resp.response,
              ...(typeof resp.id === "string" ? { id: resp.id } : {}),
            },
          });
        }
      }
    }
    return sanitized;
  }

  private requireApiKey(): void {
    if (!this.apiKey) throw new Error("Gemini API key is not configured (GOOGLE_API_KEY).");
  }

  private async request(method: string, url: string, options: RequestInit = {}): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout * 1000);
    try {
      return await fetch(url, {
        ...options,
        method,
        headers: { "Content-Type": "application/json", ...(options.headers as Record<string, string> | undefined) },
        signal: controller.signal,
      });
    } catch (exc) {
      if (exc instanceof Error && exc.name === "AbortError") {
        throw new Error("Gemini request timed out.");
      }
      throw new Error(`Could not reach Gemini: ${exc}`);
    } finally {
      clearTimeout(timer);
    }
  }

  private async apiError(response: Response, prefix: string): Promise<string> {
    const body = await response.text();
    return `${prefix} (HTTP ${response.status}): ${body}`;
  }
}
